python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll

```diff
@@ -995,6 +1017,289 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+ * will need to change type if it should be transferred from the LHS to the RHS.
+ *
+ * Note that (CodeQL modeling of) content does not have to change type on data-flow
+ * paths _inside_ the LHS, as the different allowed syntaxes here are merely a convenience.
+ * Consequently, we model all LHS sequences as tuples, which have the more precise content
+ * model, making flow to the elements more precise. If an element is a starred varibale,
+ * we will have to mutate the content type to be list content.
+ *
+ * We may for instance have
+ * ```python
+ *    (a, b) = ["a", SOURCE]  # RHS has content `ListElementContent`
+ * ```
+ * Due to the abstraction for list content, we do not know whether `SOURCE`
+ * ends up in `a` or in `b`, so we want to overapproximate and see it in both.
+ *
+ * Using wildcards we may have
+ * ```python
+ *   (a, *b) = ("a", "b", SOURCE)  # RHS has content `TupleElementContent(2)`
+ * ```
+ * Since the starred variables are always assigned (Python-)type list, `*b` will be
+ * `["b", SOURCE]`, and we will again overapproximate and assign it
+ * content corresponding to anything found in the RHS.
+ *
+ * For a precise transfer
+ * ```python
+ *    (a, b) = ("a", SOURCE)  # RHS has content `TupleElementContent(1)`
+ * ```
+ * we wish to keep the precision, so only `b` receives the tuple content at index 1.
+ *
+ * Finally, `sequence` is actually a pattern and can have a more complicated structure,
+ * such as
+ * ```python
+ *   (a, [b, *c]) = ("a", ["b", SOURCE])  # RHS has content `TupleElementContent(1); ListElementContent`
+ * ```
+ * where `a` should not receive content, but `b` and `c` should. `c` will be `[SOURCE]` so
+ * should have the content converted and transferred, while `b` should read it.
+ *
+ * The strategy for converting content type is to break the transfer up into a read step
+ * and a store step, together creating a converting transfer step.
+ * For this we need a synthetic node in the middle, which we call `TIterableElement(receiver)`.
+ * It is associated with the receiver of the transfer, because we know the receiver type (tuple) from the syntax.
+ * Since we sometimes need a converting read step (in the example above, `[b, *c]` reads the content
+ * `ListElementContent` but should have content `TupleElementContent(0)` and `TupleElementContent(0)`),
+ * we actually need a second synthetic node. A converting read step is a read step followed by a
+ * converting transfer.
+ *
+ * We can have a uniform treatment by always having two synthetic nodes and so we can view it as
+ * two stages of the same node. So we read into (or transfer to) `TIterableSequence(receiver)`,
+ * from which we take a read step to `TIterableElement(receiver)` and then a store step to `receiver`.
+ *
+ * In order to preserve precise content, we also take a flow step from `TIterableSequence(receiver)`
+ * directly to `receiver`.
+ *
+ * The strategy is then via several read-, store-, and flow steps:
```

Comment:
I think this part is unclear, and I would like to rewrite it.

```suggestion
 * To transfer content from RHS to the elements of the LHS in the expression `sequence = iterable`, we use two synthetic nodes:
 * - `TIterableSequence(sequence)` which captures the content-modeling the entire `sequence` will have 
 * (essentially just a copy of the content-modeling the RHS has)
 * - `TIterableElement(sequence)` which captures the content-modeling that will be assigned to an element. Note that
 * an empty access path means that the value we are tracking flows directly to the element.
 * 
 * Since we need to handle recursive structures on the LHS, we can have a uniform treatment by always having these 
 * two synthetic nodes and so we can view it as two stages of the same node. So we read into (or transfer to) 
 * `TIterableSequence(receiver)`, from which we take a read step to `TIterableElement(receiver)` and then a store step 
 * to `receiver`.
 *
 * This is accomplished via several read-, store-, and flow steps:
```

Reply:
Did you mean to leave out this bit "In order to preserve precise content, we also take a flow step from `TIterableSequence(receiver)` directly to `receiver`."? (Do you feel that is implied?)
Also, do you feel the two stages help? I originally wanted to make it a single branch `TIterableElement(receiver, direction)` with `direction` being `"in"` or `"out"`, but the two branches seemed much clearer end cleaner. So perhaps the two stages idea is not relevant anymore?

Reply:
> * Did you mean to leave out this bit "In order to preserve precise content, we also take a flow step from `TIterableSequence(receiver)` directly to `receiver`."? (Do you feel that is implied?)

I must admit that "In order to preserve precise content, we also take a flow step from `TIterableSequence(receiver)` directly to `receiver`." never really made sense to me.

My understanding is that the only time _Step 2_ (`TIterableSequence(sequence)` -> `sequence`) is useful, is when the content-modeling is already `TupleElementContent`. The `unpackingAssignmentFlowStep` predicate doesn't have any constraint showing this, but the only way to transfer things to the elements of the sequence is to use _Step 5_, which always requires that the content-modeling is `TupleElementContent`.

If I'm misunderstanding somewhere here, we definitely need to highlight this in some way. If my understanding is correct, we might want to add a comment to the explanation of _Step 2_, saying it's only needed to handle things (from RHS) where the content-modeling is already using `TupleElementContent`

> * Since we are here, do you feel the two stages help? I originally wanted to make it a single branch `TIterableElement(receiver, direction)` with `direction` being `"in"` or `"out"`, but the two branches seemed much clearer end cleaner. So perhaps the two stages idea is not relevant anymore?

No I'm happy about the current setup :+1:

Reply:
I ended up creating a mashup, turns out I still felt a need for justifying `TIterableSequence` 😅
Your understanding above is correct and hopefully now fully reflected in the comment.